Smart State Analysis for Hyper-V Disks

jerryk55 opened the pull request:
See the summary - "What Works and What Doesn't" at the bottom.

Add a new MiqDisk Disk Module to handle the VHDX virtual disk
format supported by the new version of Hyper-V and SCVMM.

Get Hard disk information for VMs managed
by SCVMM from the SCVMM and Hyper-V host via powershell
functions.

Read the hard disk from the Hyper-V host.

Gets the vhd, host, vhd type from the Hyper-V Host.
Opens and reads the vhd on the Hyper-V host.

This supports both VHDX files via the new VHDXDisk disk module
and the older VHD files via the existing disk modules.  Support
for communicating to Hyper-V via miq_hyperv_disk was added
to the VHD modules.

The Microsoft Provider has been extended to allow SSA scans to be
run in addition to the MiqDisk, MiqVm, and Scvmm gems/pending code.

Enhancements still required:

1) Currently Snapshots are created using VMNAME__EVM_SNAPSHOT
as the snapshot name.  An enhancement will add the date/time into the
snapshot name.

2) Caching has been implemented in the miq_hyperv_disk code which
allows the VHD and VHDX disk modules to communicate to Hyper-V
to read the files.  We need to continue to look at this caching to speed this
up as much as possible.

3) Virtual disk files residing on network-mapped drives on the Hyper-V server, seen
as starting with \\ rather than a drive letter when the inventory is taken from the server,
show up in the database with a NULL datastore ID.  These files currently fail SSA
because of this NULL ID.   https://github.com/ManageIQ/manageiq/issues/5383
describes this issue.  In addition to handling the parsing the code in MiqScvmmVm::getCfg
needs to handle this situation.  it currently does not.

4) Certain VMs seem to timeout after looping doing I/O for an inordinate amount of time while
local copies of the same VM VHD files run extremely quickly.  Further investigation is needed;
this may be related to the caching item in #2 above.

What Works And What Doesn't - 
1) SSA works on local copies of VHD and VHDX disks and their checkpoints.
2) SSA works  on remote copies of VHD and VHDX disks and their checkpoints residing on Hyper-V servers, except some larger disks may time out based on network bandwidth and possibly one other unresolved issue.
3) Hyper-V VMs with virtual disks residing on network-mapped drives starting with "\\" will not currently work.

@roliveri @Fryguy @chessbyte Please Review

By the way - an update to this old comment - this all works at this point.

Fryguy commented on lib/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,46 @@
+
```
Can you remove all these bonus blank lines at the start of every file?

jerryk55 replied:
Absatootly.  :-)

chessbyte replied:
which of the requires below is coming from util? if none, then we do not need this push.

jerryk55 replied:
miq_winrm

Fryguy commented on lib/Scvmm/miq_scvmm_parse_powershell.rb:
```diff
@@ -0,0 +1,11 @@
+
+class MiqScvmmParsePowershell
```
There is already an MiqPowershell class.  This stdout handling thing, as well as the stderr parsing thing from @bronaghs's scvmm parser code seem like they belong there instead.  Then you don't need a special class here.

jerryk55 replied:
I agree.  One of the things I want to do is move her class under lib instead of vmdb where it lives today.  it is my understanding that code under lib shouldn't be calling out to code under vmdb, although the reverse is the norm.
(One of the many "In Progress" items making this a WIP PR).

Fryguy replied:
Great! :+1:

Fryguy commented on lib/Gemfile:
```diff
@@ -5,6 +5,7 @@ source 'https://rubygems.org'
 gem "bundler", "~>1.3"
 gem "rake",    "~>10.1"
 gem "iniparse"
+gem "winrm"
```
This value is already in the vmdb/Gemfile, so will need to be removed from there if you are placing it here.  Also note that the vmdb/Gemfile value is locked to our fork of WinRM.

Fryguy commented on lib/Scvmm/miq_winrm.rb:
```diff
@@ -0,0 +1,53 @@
+  def build_uri
+    uri          = URI::HTTP.build(:port => @port, :path => "/wsman")
+    uri.hostname = @hostname
+    uri.to_s
```
You can use `:host => @hostname` now that @jrafanie has hacked it in.

bronaghs replied:
@jerryk55 
I like the MiqWinRM class. As it stands all this code is a duplication of much of EmsMicrosoft. I imagine we will want to update EmsMicrosoft to leverage this class too.

Fryguy commented on lib/Scvmm/miq_winrm.rb:
```diff
@@ -0,0 +1,53 @@
+
+require 'winrm'
```
Should this file live in a more generic location like lib/util?  I don't think it's SCVMM specific.

Fryguy commented on lib/MiqVm/MiqVm.rb:
```diff
@@ -51,6 +51,11 @@ def initialize(vmCfg, ost=nil)
+            $log.debug "MiqVm::initialize: accessing VM through SCVMM server" if $log.debug?
+            @scvmm_vm = @scvmm.get_vm(vmCfg)
+            $log.debug "MiqVm::initialize: setting @ost.miq_scvmm_vm = #{@scvmm_vm.class}" if $log.debug?
+            @vmConfig = VmConfig.new(@rhevmVm.getCfg(@ost.snapId))
```
Is `@rhevmVM` right?

jerryk55 replied:
Nope.  Just caught that myself as well.

Fryguy commented on lib/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,46 @@
+  def initialize(hyperv_host, user, pass, port = nil)
+    @hostname  = hyperv_host
+    @winrm     = MiqWinRM.new
+    winrmport  = port.nil? ? 5985 : port
```
Can this just be `port ||= 5985` and then just don't have a temporary variable?

jerryk55 replied:
Thats much nicer thanks.

Fryguy commented on lib/Scvmm/miq_scvmm_vm_ssa_info.rb:
```diff
@@ -0,0 +1,67 @@
+  end
+
+  def vm_host(vm_name)
+    host_script = <<-host_eol
```
Can you use all caps for these heredocs?  They tend to stand out better that way.  In addition, should they be left aligned?  Otherwise this send the actual leading whitespace, and I'm not sure that's intended.

bronaghs replied:
@jerryk55 
parse_single_powershell_value is being called multiply times, you could move this into a method to simplify a little.

jerryk55 replied:
I thought you meant for the name of the ruby variable, making it a constant, which would be a 'no' since we need ruby variable substitution to happen each time they are run, but  you probably mean for the name of the heredoc start/end string, which would be a 'yes'.

With regard to the leading whitespace - that certainly makes sense and I will left align the heredocs.

bronaghs commented on lib/Scvmm/miq_scvmm_vm_ssa_info.rb:
```diff
@@ -0,0 +1,67 @@
+  def vm_vhdtype(vm_name)
+    vhdtype_script = <<-vhdtype_eol
+    Get-SCVirtualHardDisk -VMMServer localhost -VM "#{vm_name}" | \
+      Select-Object -ExpandProperty VHDType
```
@jerryk55 - I don't think you need to ExpandProperty for all these simple values.

bronaghs replied:
@jerryk55 -
One thing to bear in mind…. Depending on the configuration of the VMM server, the powershell modules will reloaded into each session. This means you will need to explicitly include the line:
import-module virtualmachinemanager 
into your powershell scripts. I have not come across this myself but Nick Catling did (https://github.com/ManageIQ/manageiq/pull/2094)
You might want to test on his SCVMM server. LMK if you need the login details.

jerryk55 replied:
@bronaghs - ExpandProperty isn't necessarily *required* but it does make parsing the output eminently easier -
Without ExpandProperty the output is a blank line, the field name, a line of dashes under the field name, the field value, and another blank line.
With ExpandProperty, the output is only the field value.

I remember reading the import-module discussion as well without ever personally experiencing it.  I can certainly add the line - the only thing I can see happening is it slowing down each request slightly more.

bronaghs replied:
@jerryk55 
I agree, thats why I haven't fully embraced it. Ive looked into it and some changes can be made to the SCVMM powershell profile to dynamically load modules as they are needed but i can't imagine changing the customers box is a viable option either. So far Ive only see it in Nicks environment which is SCVMM 2012 SP1 (not R2)

chessbyte commented on lib/Scvmm/miq_scvmm_vm_ssa_info.rb:
```diff
@@ -0,0 +1,69 @@
+$LOAD_PATH.push("#{File.dirname(__FILE__)}/../util")
```
which of the requires below is coming from util?  if none, then we do not need this push.

jerryk55 replied:
miq_winrm

chessbyte commented on lib/Scvmm/test/miq_hyperv_disk_test.rb:
```diff
@@ -0,0 +1,34 @@
+$LOAD_PATH.push("#{File.dirname(__FILE__)}/..")
+
```
maybe use `require_relative` instead of updating LOAD_PATH?

chessbyte commented on lib/Scvmm/test/miq_scvmm_vm_ssa_info_test.rb:
```diff
@@ -0,0 +1,34 @@
+$LOAD_PATH.push("#{File.dirname(__FILE__)}/..")
```
maybe use `require_relative` instead of updating LOAD_PATH?

chessbyte commented on lib/disk/modules/VhdxDisk.rb:
```diff
@@ -0,0 +1,553 @@
+$LOAD_PATH.push("#{File.dirname(__FILE__)}")
+$LOAD_PATH.push("#{File.dirname(__FILE__)}/../../util")
```
maybe use `require_relative` instead of updating LOAD_PATH?

chessbyte commented on lib/disk/modules/VhdxDiskProbe.rb:
```diff
@@ -0,0 +1,29 @@
+# encoding: US-ASCII
+$LOAD_PATH.push("#{File.dirname(__FILE__)}/../../Scvmm")
```
maybe use `require_relative` instead of updating LOAD_PATH?

jerryk55 replied:
@chessbyte that can be done but is there some guidance on when to update LOAD_PATH and when to use require_relative?  I see LOAD_PATH updates throughout MIQ.....

jerryk55 commented:
@roliveri @chessbyte @Fryguy This PR has had several rounds of Rubocop cleanups and rebases as of this morning.  Review when possible please.

jerryk55 commented:
@roliveri @chessbyte @Fryguy  And again. The rebase *somehow* lost many of the commits for RuboCop cleanup that were done after the prior rebase.  They have been added back in. Hopefully we're at the end here.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager.rb:
```diff
@@ -133,4 +155,13 @@ def build_connect_params(options)
+    user = options[:user] || authentication_userid(options[:auth_type])
+    pass = options[:pass] || authentication_password(options[:auth_type])
+
+    MiqScvmmVmSSAInfo.new(vm.host.hostname, user, pass)
```
Does the current instance of `infra_manager` represent the VM's host or EMS? If it's the EMS, does the EMS's `user` and `pass` apply to the VM's host? If it's the host, should `vm.host.hostname` just be `hostname`?

jerryk55 replied:
Its definitely the host. Just hostname should work too.

roliveri replied:
If this is the host, then what's `microsoft/infra_manager/host.rb` below?

jerryk55 replied:
renamed vm_host_handle to host_handle, moved it to microsoft/infra_manager/host.rb, removed first (unnecessary) argument, and changed the first parameter to MiqScvmmVmSSAInfo.new to hostname.

jerryk55 replied:
And it turns out that hostname refers to the EMS which is the SCVMM, and we want the HyperV host here, which is vm.host.hostname, what I was originally using.  So I changed the first argument to MiqScvmmVmSSAInfo back to that.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager/host.rb:
```diff
@@ -1,44 +1,31 @@
-    _log.warn("#{$!.inspect}")
-    raise "Unexpected response returned from #{ui_lookup(:table => "ext_management_systems")}, see log for details"
-  else
+  def verify_credentials_windows(server = nil, username = nil, password = nil, _namespace = nil)
```
What happens when these options are nil?

jerryk55 replied:
I would say its not really possible for them to be nil since the Exception in line 5 should be hit first.

jerryk55 replied:
You are correct.  I'll fix.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager/vm.rb:
```diff
@@ -11,4 +13,12 @@ def self.calculate_power_state(raw_power_state)
     validate_unsupported("Migrate")
   end
+
+  def validate_smartstate_analysis
```
Does this apply to templates as well?

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager/vm.rb:
```diff
@@ -11,4 +13,12 @@ def self.calculate_power_state(raw_power_state)
+    validate_supported_check("Smartstate Analysis")
+  end
+
+  def scan_via_ems?
```
Same here - does this apply to templates?

roliveri commented on app/models/vm_or_template.rb:
```diff
@@ -1061,6 +1061,10 @@ def miq_server_proxies
       return [] if storage_id.blank? || self.repository_vm?
+    when 'RedHat'
+      return [] if storage_id.blank?
+    when 'Microsoft'
```
Type-specific versions of `miq_server_proxies` and `proxies4job` should be implemented in `microsoft/infra_manager/vm_or_template_shared/scanning.rb` - eliminating the need for type-specific checks in the common code.

jerryk55 replied:
This is an oops as well because I think I added back the line for RedHat during the merge that was there the last time but modified when redhat/infra_manager/vm_or_template_shared/scanning.rb was updated.  I'll remove the Redhat and Microsoft lines here and add the correct code in the Microsoft directory.

roliveri commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -103,9 +109,8 @@ def openDisks(diskFiles)
       dInfo.hardwareId = dtag
       dInfo.baseOnly = @ost.openParent unless mode && mode["independent"]
-      dInfo.rawDisk = @ost.rawDisk
+      dInfo.rawDisk = @ost.rawDisk # force raw disk for testing
```
Why is this comment back?

jerryk55 replied:
Probably the same as above.

roliveri commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -48,6 +48,12 @@ def initialize(vmCfg, ost = nil)
       @vmConfig = VmConfig.new(getCfg(@ost.snapId))
       $log.debug "MiqVm::initialize: @vmConfig.getHash = #{@vmConfig.getHash.inspect}"
       $log.debug "MiqVm::initialize: @vmConfig.getDiskFileHash = #{@vmConfig.getDiskFileHash.inspect}"
+    elsif @ost.miq_hyperv
```
Might as well add another `# TODO: move this to` comment before this `elsif` as well.

jerryk55 replied:
Will do.

roliveri commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -64,7 +70,7 @@ def vmDisks
 
   def openDisks(diskFiles)
-    pVolumes = []
+    pVolumes = Array.new
```
Why did this change?

jerryk55 replied:
Because RuboCop said to change it.  It likes [] instead of Array.new and {} instead of Hash.new.

jerryk55 replied:
This must have been a change that got unchanged somehow with the rebase.  I'll change it back.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager/vm_or_template_shared/scanning.rb:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        ost.miq_scvmm  = ost.miq_hyperv = hyperv_config
+        ost.miq_vm     = ost.fileName = vm_name
```
Is this actually connecting to anything?

jerryk55 replied:
Name is copied from the Redhat and VMware code which both do the same thing.  No - it doesn't connect within connect_to_ems - so they all should be changed - it sets up the OpenStruct data in order to allow it to connect, then in the calling code in perform_metadata_scan it uses the oat in the call to MiqScvmmVm in this case, MiqVM in the VMware case, and MiqRhevmVm in the Redhat case.

roliveri replied:
No, the Redhat and VMware versions actually connect, they always have.
VMware: 
```
ost.miqVim = MiqFaultTolerantVim.new(...
```
Redhat:
```
ost.miqRhevm = ext_management_system.rhevm_inventory
```
and before that, the Redhat version used to connect through `oVirt` directly.

roliveri commented on gems/pending/MiqVm/miq_scvmm_vm.rb:
```diff
@@ -0,0 +1,85 @@
+require 'Scvmm/miq_scvmm_vm_ssa_info'
+
+class MiqScvmmVm < MiqVm
+  def openDisks(disk_files)
```
Can't you just use the `openDisks` method in the superclass? Is it only because of the `init_disk_info` usage?
It would be better if the disk, partition, and volume handling remained common.

roliveri commented on gems/pending/MiqVm/miq_scvmm_vm.rb:
```diff
@@ -0,0 +1,85 @@
+
+      part = disk.getPartitions
+      if part.empty?
+        part_volumes << disk
```
This should be `physical_volume` not `part_volume`.

roliveri commented on gems/pending/MiqVm/miq_scvmm_vm.rb:
```diff
@@ -0,0 +1,85 @@
+        @diskInitErrors[disk_file] = err.to_s
+        next
+      end
+
```
What happened to the setting of `@wholeDisks`? If that's not set, the `umount` may not work properly.
This is another reason I'd prefer this aspect of the code remain common.

jerryk55 commented:
@roliveri All changes based on your comments except for the ones for miq_server_proxies and proxies4job have been made.  openDisks has been removed from miq_sccvm_vm.rb and a minor change in MiqVm's openDisks allows this to work properly.  We can still discuss the above if you would like.  If you have other comments I will address them as well.  For now I have re-pushed this PR.

jerryk55 commented:
@roliveri another commit added - 
1) A bug was found causing VMware SSA to rail with "VM not found" caused by a change in server_smart_proxy.rb.  
2) miq_hyperv_disk.rb changed 4K blocks read in cache from 16 to 4 to try to speed up read processing.
3) Another RuboCopy cleanup.

jerryk55 commented:
Adding 2 fixes relating to allowing SSA for Hyper-V to succeed.  1) Increase the miq_hyperv_disk buffer cache from 200 to 300 entries.  2) Add a per-provider timeout adjustment for the VmScan timeout in the current_job_timeout method.  It is set to 1 by default but set to 4 for Microsoft so the slow network reads for Hyperv SSA doesn't cause the scans to time out.

@roliveri @Fryguy @chessbyte If you can finish the review we can try to merge this for the sprint next week.  Thanks much.

roliveri commented:
@jerryk55 - I'm going over the code again. In the meantime, you need to look into resolving the test failures.

roliveri commented on gems/pending/MiqVm/miq_scvmm_vm.rb:
```diff
@@ -0,0 +1,47 @@
+      user = @ost.miq_hyperv[:domain] + "\\" + @ost.miq_hyperv[:user]
+    end
+    password          = @ost.miq_hyperv[:password]
+    scvmm_info_handle = MiqScvmmVmSSAInfo.new(host, user, password, port)
```
This is contrary to the pattern established by the other provider types. Typically, the connection to the API handle is established outside of this object and passed in to initialize, instead of establishing it in initialize. The connection should be established in `connect_to_ems` and passed in via the `ost`. This decouples the establishment of the connection from the instantiation the `MiqScvmmVm`, making it more suitable for generalized use.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager/vm_or_template_shared/scanning.rb:
```diff
@@ -0,0 +1,79 @@
+
+  # Moved from MIQExtract.rb
+  # TODO: Should this be in the ems?
+  def setup_for_ems_connection(vm_name, ost)
```
I wasn't taking issue with the name. This should be `connect_to_ems`, but it should actually connect here, not in `MiqScvmmVm`. The connection should be passed into `MiqScvmmVm` via the `ost`.

jerryk55 replied:
Yeah I don't know why I wasn't getting what you were saying but that makes perfect sense.  I'll make the required changes here and in miq_scvmm_vm.

roliveri commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -48,6 +48,13 @@ def initialize(vmCfg, ost = nil)
+    # TODO: move this to miq_scvmm_vm
+    elsif @ost.miq_hyperv
+      $log.debug "MiqVm::initialize: accessing VM through HyperV server" if $log.debug?
+      @scvmm_vm = @ost.miq_vm
```
Does `@scvmm_vm` get used anywhere?

jerryk55 replied:
I think it was previously but can be removed now.

roliveri commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -48,6 +48,13 @@ def initialize(vmCfg, ost = nil)
+    elsif @ost.miq_hyperv
+      $log.debug "MiqVm::initialize: accessing VM through HyperV server" if $log.debug?
+      @scvmm_vm = @ost.miq_vm
+      @ost.miq_scvmm_vm = @scvmm_vm
```
Same here, does `@ost.miq_scvmm_vm` get used antwhere?

jerryk55 commented:
@roliveri Made recommended changes with regard to connect_to_ems and passing the  connection to MiqVM so getCfg doesn't have to set up the connection all over again.  Also fixed up the timeout issue that was causing the tests to fail (Hopefully!).

jerryk55 commented:
@Fryguy @roliveri @chessbyte Any chance of this getting a final review and merged before the break?  I'm still looking at speed enhancements but the functionality is working as is.

roliveri commented:
@jerryk55 - I'll go through the code one more time today. In the meantime, given there are some changes to common SSA code, have/can you test SSA for some other providers - just in case there's some unforeseen breakage?

jerryk55 commented:
@roliveri I've tested VMware SSA with these changes.  There was a bug that I found if you recall - the "VM not Found" issue.  I will test again with the latest.

roliveri commented on app/models/job.rb:
```diff
@@ -171,7 +171,7 @@ def self.check_jobs_for_timeout
         .where("zone is null or zone = ?", MiqServer.my_zone)
         .each do |job|
-          next unless job.updated_on < job.current_job_timeout.seconds.ago
+          next unless job.updated_on < job.current_job_timeout(timeout_adjustment(job)).seconds.ago
```
It looks like `current_job_timeout` always returns `DEFAULT_TIMEOUT`, so how does this adjust the timeout?

roliveri replied:
Ah, I see, overridden in the subclass.

roliveri commented on app/models/manageiq/providers/microsoft/infra_manager.rb:
```diff
@@ -133,4 +155,13 @@ def build_connect_params(options)
     connect_params
   end
+
+  def host_handle(vm, options)
```
So, if we want to get the handle associated with the VM's host - `vm.host` - then this method should be defined in the `vm.host` object. Then line 108 above becomes: `host_handle = vm.host.host_handle(options)`.

In addition, doesn't the authentication information come from what's defined in `vm.host`? If so, then there should be no need to pass in `options`, resulting in: `host_handle = vm.host.host_handle`.

jerryk55 replied:
@roliveri the second paragraph above is a little nebulous.  The auth info is defined in the host where? I thought maybe that "options" was known to the host object but unless I pass it in like
host_handle = vm.host.host_handle(options) it doesn't know the variable.

roliveri replied:
@jerryk55 - if `vm.host` is the HyperV host of the VM, then that should be the `Host` model object for the host. I'm assuming the relationship is similar to that of VMware - where SCVMM is like vSphere and HyperV is like ESX. So, if the credentials for HyperV hosts can be different than those of their associated SCVMM, we need to maintain credentials for each HyperV host - like we do for ESX. Then each host `has_many` authentications, just as the SCVMM EMS does.

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,142 @@
+    @file_size          = file_size.to_i
+    @end_byte_addr      = @file_size - 1
+    # @size_in_blocks     = @file_size / @block_size
+    @size_in_blocks     = @file_size / @block_size + 1
```
I think this should be something like:
```
@size_in_blocks, rem  = @file_size.divmod(@block_size)
@size_in_blocks += 1 if rem > 0
```

jerryk55 replied:
absolutely correct.

jerryk55 commented:
@roliveri I've made the one change in miq_hyperv_disk regarding the calculation of the number of blocks.  I've also fixed up the host_handle code, moving from microsoft/inframanager.rb to microsoft/inframanager/host.rb with the changes we've discussed.

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,176 @@
+        @cache_hits[start_sector] += 1
+      else
+        @cache_hits[start_sector] = 1
+      end
```
nitpick: Setting a default value of `0` for the `@cache_hits` hash in `initialize` would eliminate the need for this condition, replacing the entire condition with: `@cache_hits[start_sector] += 1`

No need to change this unless other changes are also needed.

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,176 @@
+      @cache_misses[start_sector] += 1
+    else
+      @cache_misses[start_sector] = 1
+    end
```
Same as above for `@cache_hits`

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,176 @@
+    encoded_data = @parser.output_to_attribute(@winrm.run_powershell_script(read_script))
+    t2           = Time.now.current
+    buffer       = MemoryBuffer.create(@block_size * number_sectors)
+    buffer       = ""
```
Is this throwing away the buffer that was allocated the line before?

jerryk55 replied:
Ok, so that is not the intent, but does appear to be what it is doing. The intent is to fill the buffer with the decoded date in the line below.  Let me look at what is really happening.

jerryk55 commented:
@roliveri I have fixed that issue with the MemoryBuffer in miq_hyperv_disk.rb as well as the initialization of the cache_hits and cache_misses hashes.  Also rebased and repushed due to a conflict.

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,171 @@
+    @file_size   = 0
+    @block_cache = LruHash.new(DEF_BLOCK_CACHE_SIZE)
+    @cache_hits  = {}
+    @cache_misses = {}
```
Sorry, I wasn't clear. You can set a default value for a hash when it's instantiated. The previous 2 lines should be:
```
@cache_hits  = Hash.new(0)
@cache_misses = Hash.new(0)
```
Then, the initialization code in `open` can be removed.

jerryk55 replied:
Oh.  cool.  Ok.

On Thu, Dec 17, 2015 at 4:00 PM, Richard Oliveri <notifications@github.com>
wrote:

> In gems/pending/Scvmm/miq_hyperv_disk.rb
> <https://github.com/ManageIQ/manageiq/pull/2428#discussion_r47960304>:
>
> > +  DEBUG_CACHE_STATS    = false
> > +
> > +  attr_reader :hostname, :virtual_disk, :file_offset, :file_size, :parser, :vm_name, :temp_snapshot_name
> > +
> > +  def initialize(hyperv_host, user, pass, port = nil)
> > +    @hostname  = hyperv_host
> > +    @winrm     = MiqWinRM.new
> > +    port ||= 5985
> > +    options = {:port => port, :user => user, :pass => pass, :hostname => @hostname}
> > +    @connection  = @winrm.connect(options)
> > +    @parser      = MiqScvmmParsePowershell.new
> > +    @block_size  = 4096
> > +    @file_size   = 0
> > +    @block_cache = LruHash.new(DEF_BLOCK_CACHE_SIZE)
> > +    @cache_hits  = {}
> > +    @cache_misses = {}
>
> Sorry, I wasn't clear. You can set a default value for a hash when it's
> instantiated. The previous 2 lines should be:
>
> @cache_hits  = Hash.new(0)
> @cache_misses = Hash.new(0)
>
> Then, the initialization code in open can be removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ManageIQ/manageiq/pull/2428/files#r47960304>.
>



-- 

-- 
Jerry Keselman
Red Hat Virtualization R&Djerryk@redhat.com | p. 212 530 7873

roliveri commented on gems/pending/Scvmm/miq_hyperv_disk.rb:
```diff
@@ -0,0 +1,171 @@
+    (0..@size_in_blocks - 1).each do |block|
+      @cache_hits[block]   = 0
+      @cache_misses[block] = 0
+    end
```
This initialization block can be removed, given the changes described in `initialize`.

jerryk55 commented:
@roliveri cache hit & miss hash initialization code cleaned up.  Problem with MemoryBuffer.create buffer getting thrown away fixed. And all pushed.

roliveri commented on gems/pending/fs/MiqFS/FsProbe.rb:
```diff
@@ -14,6 +14,7 @@ module FsProbe
   PROBE_FILES.unshift("Ext4Probe")     if PROBE_FILES.delete("Ext4Probe")
   PROBE_FILES.unshift("Ext3Probe")     if PROBE_FILES.delete("Ext3Probe")
   PROBE_FILES.unshift("NTFSProbe")     if PROBE_FILES.delete("NTFSProbe")
+  PROBE_FILES.unshift("XFSProbe")      if PROBE_FILES.delete("XFSProbe")
```
What does this change have to do with this PR?

jerryk55 replied:
I am under the impression that an early version of the WIP PR, when rebased, somehow ended up deleting this like that was added (or should have been added?) by the XFS PR.  I added it back thinking that it was inadvertent.  If you want it deleted and added under a separate PR let me know.

roliveri replied:
I don't understand. The version on master doesn't contain this line, so if it were inadvertently removed, it had to be merged.  If you're sure the line belongs there, then leave it, but it does worry me that it was deleted in the first place.

roliveri commented on gems/pending/metadata/linux/LinuxSystemd.rb:
```diff
@@ -32,7 +32,8 @@ def parse_systemd
     def files(unit_extension)
-      (SYSTEM_DIRS + USER_DIRS).flat_map do |dir|
+      dirs = my_dirs
+      dirs.flat_map do |dir|
```
What do the changes in this file have to do with this PR?

jerryk55 replied:
I think I had actually accidentally deleted it in an earlier version of the WIP PR during a rebase and then put it back in.  That's my story and I"m sticking to it.  If you would like it removed let me know.

jerryk55 replied:
I came across this SystemD issue when parsing a CentOS VM for this PR.  But you are correct, these should probably be added under a separate PR.  I can do so if you like.

roliveri replied:
Yes, this should be in it's own PR. You should probably open an issue describing the problem as well.

jerryk55 commented:
@Fryguy @chessbyte Two commits were removed from the PR which were somewhat orthogonal to the functionality and they will be added as separate PRs.  I have no other comments on this PR from @roliveri.  As far as I know this is ready to go (speed notwithstanding).

roliveri commented:
Looks good to me.
@chessbyte @Fryguy - feel free to merge when you think it's ready.

Fryguy commented on gems/pending/MiqVm/MiqVm.rb:
```diff
@@ -300,7 +306,7 @@ def format(event)
     if rt.guestOS == "Linux"
       puts "\n\t\t*** /etc/fstab contents:"
-      rt.fileOpen("/etc/fstab", &:read).each_line do |fstl|
+      rt.fileOpen("/etc/fstab") { |fo| fo.read }.each_line do |fstl|
```
Why was this changed?  Not that it matters, but they are equivalent.

jerryk55 replied:
Rubocop.

Fryguy replied:
Interesting...I would have though Rubocop would go the other way.

Fryguy replied:
Actually Rubocop *did* go the other way...see the Rubocop output above (MiqVm.rb Line 309)

jerryk55 replied:
Oh.  Possibly this was a rebase conflict in which I chose the wrong side of the diff - because rest assured I wouldn't have come up with that line on my own... its probably the old code.  I'll change it back.

Fryguy commented:
I'm pretty sure @chessbyte worked to get rid of LOAD_PATH manipulation.  Can we remove the LOAD_PATH manipulation and just do full requires?

jrafanie commented on gems/pending/disk/modules/VhdxDisk.rb:
```diff
@@ -0,0 +1,634 @@
+  def page_83_data(_offset)
+    page_83_metadata = VHDX_PAGE_83_DATA.decode(@vhdx_file.read(SIZEOF_VHDX_PAGE_83_DATA))
+    @page_83_data    = page_83_metadata['page_83_data']
+    $log.debug "Page 83 Data #{@page_83_data}"
```
@jerryk55 I can't tell if any of these debug messages would contain lots of data or involve complex operations.  Even though you're doing .debug here, the String is still constructed to be passed to `$log.debug`, just to be a no-op for the default info log level.  If any of these debug messages are big or complex, we should use `$log.debug { "Page 83 Data #{@page_83_data}" }` style so the string isn't constructed unless it's debug.  cc @matthewd

jerryk55 replied:
@jrafanie in the three calls to $log.debug you cite above, the definitions of the data in question in each case is in the preceding line, so no, there isn't a lot of data or complex operations - one is true or false, one is an integer, and one is a string.  I can make the changes if you like, but are you suggesting that *all* calls to $log.debug where there are variable substitutions in the string should use this methodology?

jrafanie replied:
No, sorry, not this specific one, nor should all use the block notation.  For minor string building, the cost of constructing the String is probably close enough as constructing a closure for the block notation.

My browser counts 55 instances of `log.debug` in the changes... I didn't go through all of them.  Some are obviously not a big deal... I'm asking if any of these strings we're building are large or involves complicated operations.  Any could be large or expensive, it's good to make them use block notation to avoid work we're just going to throw away at info logging.

jerryk55 replied:
@jrafanie I'd be interested to see your data since I can only find about half that number of calls.  I did find one case of err.backtrace.join(*\n") in a log.debug call that could be enclosed in brackets but all the others are cases of simple variable substitution.  Also note that of the 24 log.debug calls using err.backtrace.join in the entire source tree including this PR only 5 are enclosed in brackets so if this is indeed an issue I can add a PR to rectify the situation.  Let me know if this is necessary.

jrafanie replied:
Yeah, the numbers are what the browser is finding for `log.debug` which includes unchanged source lines surrounding your changes so there's definitely less than 50 matches.  I'm fine with merging this as is :+1: , it's merely a  FYI comment to say that if the log message is relatively big or intensive to build, we might want to use bracket notation to avoid that cost in the default case, info.

jrafanie commented on gems/pending/disk/modules/VhdxDisk.rb:
```diff
@@ -0,0 +1,634 @@
+    @vhdx_file
+  end
+
+  def d_read(pos, len)
```
It would be nice to have targeted tests for complicated methods like this.

jrafanie replied:
I wonder if we could run VCR like tests of this method by running a full end to end test of some scvmm disks, recording the inputs and outputs as test cases.  It might help better understand the conditionals in this method.

jerryk55 replied:
I would suppose that is a possibility.  I'm not sure if we have anything like that for any of the disk formats we already support, by the way.

roliveri replied:
@jrafanie @jerryk55 - I hope to enable testing things like this in a general way very soon - once we're able to deploy  the camcorder FS. Then, we should be able to write tests for all disk file formats.

jerryk55 replied:
And in the meantime I would think that the lack of the general test would not be a blocker to merging.

jerryk55 commented:
@Fryguy yes I'll look into that in VhdxDisk.rb, VhdxDiskProbe.rb, MSVSDiskProbe.rb, miq_hyperv_disk.rb, miq_scvmm_vm_ssa_info.rb, as well as host.rb and infra_manager.rb in the microsoft provider.  I do see it still out there in a few spots besides this - are these places that were left over, or places it was added back by other PRs? For instance under app/models host.rb, filesystem.rb, miq_server/server_smart_proxy.rb.

chessbyte commented:
@jerryk55 In general, this PR has several challenges.

1. Too much new code vs little/no new specs.
2. Too many unrelated changes or changes that should be split out into separate PRs.  (eg new VHDx disk format).
3. Too many changes/additions for the reviewer to absorb in one PR.

Lets discuss further in 2016.

Thanks,
Oleg

jerryk55 commented:
@chessbyte okie dokie.  Yes I've always thought that VhdxDisk could be its own PR, and the HyperVDisk methods could also stand on their own.  Thanks.

jerryk55 commented:
@roliveri @chessbyte @Fryguy @jrafanie This PR has been amended to pull the VhdxDisk module out as discussed.  I will try to reword the log message of one or two of the commit messages but otherwise it is ready for any further review and merging when ready.

chessbyte commented on app/models/manageiq/providers/microsoft/infra_manager.rb:
```diff
@@ -1,4 +1,4 @@
-$LOAD_PATH << File.join(GEMS_PENDING_ROOT, "Scvmm")
+# encoding: US-ASCII
```
why is this line being added?

jerryk55 replied:
Because I was under the mistaken impression that was standard practice in our code, but now I see that not to be the case.

jerryk55 commented:
@chessbyte Removed the above extraneous encoding comments and re-pushed.
